Photon: Auto-generate additional srcset options

kraftbj opened the pull request:
WIP
See #2918

~~~~This PR automatically generates additional srcset options based on % of the original sized images. ~~~~

This PR generates additional srcset options based on a variety of factors.

Outstanding questions:
* What about $max_srcset_image_width? It is 1600px by default with a filter.
* Why does ^^ exist? What is the rationale for excluding srcset sources over 1600?
* Custom crops? Are we handling these gracefully or do we need some checking?
* Unit tests. I couldn't figure out how to mock it in tests while in flight, so just dove in.
* Per @lancewillett's [comment](https://github.com/Automattic/jetpack/issues/2918#issuecomment-215109837), would using `$content_width` provide more flexibility?

ChrisBegley commented:
This looks great! I just tested this and it appears to be working as expected on images that are inserted into a post. It doesn't seem to be handling custom crops though. For example, I have 80x80 images on my sidebar, and the images are distorted because they aren't being cropped anymore: 

`<img width="80" height="80" src="http://i1.wp.com/techyourworld.com/wordpress/wp-content/uploads/2014/02/Superman-Man-of-Steel-Logo-Imgur.jpeg?resize=80%2C80" class="attachment-recent-thumbnail default-featured-img" alt="Superman-Man-of-Steel-Logo-Imgur" srcset="http://i1.wp.com/techyourworld.com/wordpress/wp-content/uploads/2014/02/Superman-Man-of-Steel-Logo-Imgur.jpeg?w=1728 1728w, http://i1.wp.com/techyourworld.com/wordpress/wp-content/uploads/2014/02/Superman-Man-of-Steel-Logo-Imgur.jpeg?w=1536 1536w, http://i1.wp.com/techyourworld.com/wordpress/wp-content/uploads/2014/02/Superman-Man-of-Steel-Logo-Imgur.jpeg?w=1344 1344w, http://i1.wp.com/techyourworld.com/wordpress/wp-content/uploads/2014/02/Superman-Man-of-Steel-Logo-Imgur.jpeg?w=1152 1152w, http://i1.wp.com/techyourworld.com/wordpress/wp-content/uploads/2014/02/Superman-Man-of-Steel-Logo-Imgur.jpeg?w=960 960w, http://i1.wp.com/techyourworld.com/wordpress/wp-content/uploads/2014/02/Superman-Man-of-Steel-Logo-Imgur.jpeg?w=768 768w, http://i1.wp.com/techyourworld.com/wordpress/wp-content/uploads/2014/02/Superman-Man-of-Steel-Logo-Imgur.jpeg?w=576 576w" sizes="(max-width: 80px) 100vw, 80px">`

The auto-generation is awesome, but if we wanted more control would it be possible to output our own srcset sources?

kraftbj commented:
@ChrisBegley - Thanks for the testing and feedback. I believe you're right that I need to add checking to make sure the srcset proportions match the proportions of the original image if we're going to include it. The commit I just pushed do *not* include that yet.

> we wanted more control would it be possible to output our own srcset sources?

Right now, with the PR as it stands, you can use Core's filter before priority 10 to amend the srcset values that Core natively provides, then use the filter in the PR to "short-circuit" the auto-creation.

kraftbj commented:
In https://github.com/Automattic/jetpack/pull/3748/commits/14941748804001625a9c31a8f92cd946de6b06e3 : 

 ### Add a `sizes` array filter.
Original patch props @aduth 
Right now, if there is not something already filtering the sizes array and we have a `content_width`, we're using that to max out the sizes array.

Since we've been debating this point back and forth, you can disable this filter by adding a `?photontest=nosize` query string to the URL for easy testing.

If the content_width is not set, we are using the original Core behavior for now. We can/should add another maximum, but I ran out of time today.

### srcset improvements
If there is a content_width set, we will attempt to generate srcset values of 2x and 3x content_width, assuming that'll cover the maximum displayed size on the vast majority of HiDPI devices. (Need to test: What happens when using a 3.5x or 4x device? Does it still grab the full size? Is that okay)

If there is no content_width, we'll generate a few options at 80%, 60%, and 40% of the original image size.

### Testing needed
Still needs a ton of testing, not ready for merge. Just to keep the conversation moving.

aduth commented on class.photon.php:
```diff
@@ -60,6 +60,9 @@ private function setup() {
 
 		// Responsive image srcset substitution
 		add_filter( 'wp_calculate_image_srcset', array( $this, 'filter_srcset_array' ), 10, 4 );
+		if ( ! has_filter( 'wp_calculate_image_sizes') ) { // To allow themes to fully control the sizes array.
```
Could we achieve the same effect by setting a lower priority to our filter, since if a theme had defined their own, it would be run and filtered later (assuming default priority of 10)?

kraftbj replied:
I'm open to it. I don't have strong opinions on how we do it, as long as we let theme authors be smarter than we can be globally. I'm trying to think what kind of use case would modifying it before a theme gets it would cause problems; nothing yet.

aduth commented on class.photon.php:
```diff
@@ -646,9 +650,86 @@ public function filter_srcset_array( $sources, $size_array, $image_src, $image_m
+					// If a new width would be within 100 pixes of an existing one, skip.
+					if ( abs( $currentwidth - $newwidth ) < 50 ) {
+						$usewidth = false;
+						break;
```
It's an uncommon syntax that may harm readability, but there's a case for [specifying a numeric argument for `continue`](http://php.net/manual/en/control-structures.continue.php) here:

```php
foreach ( $multipliers as $multiplier ) {
	$newwidth = round( $base * $multiplier );
	foreach ( $currentwidths as $currentwidth ) {
		// If a new width would be within 100 pixes of an existing one, skip.
		if ( abs( $currentwidth - $newwidth ) < 50 ) {
			continue 2;
		}
	}

	$newsources[ $newwidth ] = array(
		'url'         => jetpack_photon_url( $url, array( 'w' => $newwidth ) ),
		'descriptor'  => 'w',
		'value'       => $newwidth,
	);
} // foreach ( $multipliers as $multiplier )
```

kraftbj replied:
Likewise, I dig it. I'll make the change and inline comment it to help readability.

josephscott commented:
@ChrisBegley can you provide detailed steps on how to reproduce ( and make sure that you are using the current version of the patch ).  I've tried out cropped images ( via image edit in wp-admin ) in posts and so far they appear normal.  If there is something we are missing I want to make sure that we get reproduced so we can address it.

kraftbj commented:
I think I have it duplicated. Used this in a plugin/theme functions: 
```
add_image_size('kraft-hard-crop', 200, 200, true);
add_filter('image_size_names_choose', 'my_image_sizes');
function my_image_sizes($sizes) {
$addsizes = array(
"kraft-hard-crop" => __( "Kraft Hard Crop")
);
$newsizes = array_merge($sizes, $addsizes);
return $newsizes;
}
```
then inserted the "Kraft Hard Crop" image via wp-admin. The resulting srcset include hard resize Photon args from the core produced sizes (as expected) while the new ones from this PR are soft-cropped, e.g. `https://i0.wp.com/kraftbj.wpsandbox.me/wp-content/uploads/2016/05/DSC_1387-1.jpg?w=4439&ssl=1 4439w, https://i0.wp.com/kraftbj.wpsandbox.me/wp-content/uploads/2016/05/DSC_1387-1.jpg?w=3329&ssl=1 3329w, https://i0.wp.com/kraftbj.wpsandbox.me/wp-content/uploads/2016/05/DSC_1387-1.jpg?w=2220&ssl=1 2220w"`

ChrisBegley commented:
@josephscott oops, I think I misunderstood what @kraftbj meant by "custom crops". I haven't tried inserting images that I cropped with WordPress into posts. I was referring to custom thumbnail images that I have in my theme that use crops. Like this:

`  add_image_size( 'recent-thumbnail',     80,    80, true );`

josephscott commented:
@ChrisBegley that makes sense, when I saw "crop" my mind when immediately to the crop tool in the image editor of wp-admin.

Just so I make sure I've got all of this right, with this patch you end up with http://i1.wp.com/techyourworld.com/wordpress/wp-content/uploads/2014/02/Superman-Man-of-Steel-Logo-Imgur.jpeg?resize=80%2C80 - what does pre-patch crop version look like?

ChrisBegley commented:
@josephscott yes, that's exactly right. Pre-patch it was using `resize`, so I got the crop and aspect ratio that I expected. Now it's using `w`, so it's not cropping the thumbnail images.

So I used to get this:
http://i1.wp.com/techyourworld.com/wordpress/wp-content/uploads/2014/02/Superman-Man-of-Steel-Logo-Imgur.jpeg?resize=80,80

Now I get something like this:
http://i1.wp.com/techyourworld.com/wordpress/wp-content/uploads/2014/02/Superman-Man-of-Steel-Logo-Imgur.jpeg?w=576

It's not longer being resized and cropped to the dimensions defined in my theme:
`add_image_size( 'recent-thumbnail', 80, 80, true );`

kraftbj commented:
70032ea provides some support for hard-cropped image sizes. @ChrisBegley, can you give it a whirl?

We probably still need to add checks, across the board, to not insert a srcset that is bigger than the full-size image. At this point, Photon will return the original size image in those cases, but not ideal.

ChrisBegley commented:
@kraftbj looks great on my end! It automatically added in 2x and 3x versions too, which is awesome. Let me know if there's anything else you want me to take a look at.

kraftbj commented:
Thanks @ChrisBegley! I added a check to make sure we're not adding entries larger than the full width image.

Before, it was doing things like for an original image size of 150x150:
```
https://i0.wp.com/kraftbj.wpsandbox.me/wp-content/uploads/2016/05/150150.png?w=150&ssl=1 150w,
https://i0.wp.com/kraftbj.wpsandbox.me/wp-content/uploads/2016/05/150150.png?resize=100%2C100&ssl=1 100w,
https://i0.wp.com/kraftbj.wpsandbox.me/wp-content/uploads/2016/05/150150.png?w=1208&ssl=1 1208w,
https://i0.wp.com/kraftbj.wpsandbox.me/wp-content/uploads/2016/05/150150.png?w=1812&ssl=1 1812w
```

Now, it'll abort for any variants where the new size is larger than the original, based on width.

kraftbj commented:
Marking this for 4.0.4 as this behavior is a bug (making Photon less performant than previously).

jeherve commented:
Made 2 minor changes, but otherwise this looks good to merge to me (and @kraftbj told me it was okay :) )

kraftbj commented:
Do we want to squash this down until a couple of atomic commits?

samhotchkiss commented:
Squashing would be great, @kraftbj -- also, Travis needs to pass :)

kraftbj commented:
Squashed, rebased against master to pick up Travis improvements (this PR was still running the double-set of Travis builds). No tests were failing, just cancelled them due to yesterday's Travis backlog. Will merge/cherry-pick once tests pass.
